File: src/backend/estoque/src/model/estoque_grade_model.py
from src import db
from src.model.produto_model import ProdutoModel
from src.model.estoque_cor_model import EstoqueCorModel
from src.model.estoque_tamanho_model import EstoqueTamanhoModel
from src.model.estoque_sabor_model import EstoqueSaborModel
from src.model.estoque_marca_model import EstoqueMarcaModel


class EstoqueGradeModel(db.Model):
    __tablename__ = 'estoque_grade'

    id = db.Column(db.Integer, primary_key=True)
    codigo = db.Column(db.String(50))
    quantidade = db.Column(db.Float)
    id_produto = db.Column(db.Integer, db.ForeignKey('produto.id'))
    id_estoque_cor = db.Column(db.Integer, db.ForeignKey('estoque_cor.id'))
    id_estoque_tamanho = db.Column(db.Integer, db.ForeignKey('estoque_tamanho.id'))
    id_estoque_sabor = db.Column(db.Integer, db.ForeignKey('estoque_sabor.id'))
    id_estoque_marca = db.Column(db.Integer, db.ForeignKey('estoque_marca.id'))

    produto_model = db.relationship('ProdutoModel', foreign_keys=[id_produto])
    estoque_cor_model = db.relationship('EstoqueCorModel', foreign_keys=[id_estoque_cor])
    estoque_tamanho_model = db.relationship('EstoqueTamanhoModel', foreign_keys=[id_estoque_tamanho])
    estoque_sabor_model = db.relationship('EstoqueSaborModel', foreign_keys=[id_estoque_sabor])
    estoque_marca_model = db.relationship('EstoqueMarcaModel', foreign_keys=[id_estoque_marca])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_produto = data.get('idProduto')
        self.id_estoque_marca = data.get('idEstoqueMarca')
        self.id_estoque_sabor = data.get('idEstoqueSabor')
        self.id_estoque_tamanho = data.get('idEstoqueTamanho')
        self.id_estoque_cor = data.get('idEstoqueCor')
        self.codigo = data.get('codigo')
        self.quantidade = data.get('quantidade')

    def serialize(self):
        return {
            'id': self.id,
            'idProduto': self.id_produto,
            'idEstoqueMarca': self.id_estoque_marca,
            'idEstoqueSabor': self.id_estoque_sabor,
            'idEstoqueTamanho': self.id_estoque_tamanho,
            'idEstoqueCor': self.id_estoque_cor,
            'codigo': self.codigo,
            'quantidade': self.quantidade,
            'produtoModel': self.produto_model.serialize() if self.produto_model else None,
            'estoqueCorModel': self.estoque_cor_model.serialize() if self.estoque_cor_model else None,
            'estoqueTamanhoModel': self.estoque_tamanho_model.serialize() if self.estoque_tamanho_model else None,
            'estoqueSaborModel': self.estoque_sabor_model.serialize() if self.estoque_sabor_model else None,
            'estoqueMarcaModel': self.estoque_marca_model.serialize() if self.estoque_marca_model else None,
        }

File: frontend/estoque/lib/app/data/model/estoque_grade_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:estoque/app/data/model/model_imports.dart';

class EstoqueGradeModel {
	int? id;
	int? idProduto;
	int? idEstoqueMarca;
	int? idEstoqueSabor;
	int? idEstoqueTamanho;
	int? idEstoqueCor;
	String? codigo;
	double? quantidade;
	ProdutoModel? produtoModel;
	EstoqueCorModel? estoqueCorModel;
	EstoqueTamanhoModel? estoqueTamanhoModel;
	EstoqueSaborModel? estoqueSaborModel;
	EstoqueMarcaModel? estoqueMarcaModel;

	EstoqueGradeModel({
		this.id,
		this.idProduto,
		this.idEstoqueMarca,
		this.idEstoqueSabor,
		this.idEstoqueTamanho,
		this.idEstoqueCor,
		this.codigo,
		this.quantidade,
		this.produtoModel,
		this.estoqueCorModel,
		this.estoqueTamanhoModel,
		this.estoqueSaborModel,
		this.estoqueMarcaModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'codigo',
		'quantidade',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Codigo',
		'Quantidade',
	];

	EstoqueGradeModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idProduto = jsonData['idProduto'];
		idEstoqueMarca = jsonData['idEstoqueMarca'];
		idEstoqueSabor = jsonData['idEstoqueSabor'];
		idEstoqueTamanho = jsonData['idEstoqueTamanho'];
		idEstoqueCor = jsonData['idEstoqueCor'];
		codigo = jsonData['codigo'];
		quantidade = jsonData['quantidade']?.toDouble();
		produtoModel = jsonData['produtoModel'] == null ? ProdutoModel() : ProdutoModel.fromJson(jsonData['produtoModel']);
		estoqueCorModel = jsonData['estoqueCorModel'] == null ? EstoqueCorModel() : EstoqueCorModel.fromJson(jsonData['estoqueCorModel']);
		estoqueTamanhoModel = jsonData['estoqueTamanhoModel'] == null ? EstoqueTamanhoModel() : EstoqueTamanhoModel.fromJson(jsonData['estoqueTamanhoModel']);
		estoqueSaborModel = jsonData['estoqueSaborModel'] == null ? EstoqueSaborModel() : EstoqueSaborModel.fromJson(jsonData['estoqueSaborModel']);
		estoqueMarcaModel = jsonData['estoqueMarcaModel'] == null ? EstoqueMarcaModel() : EstoqueMarcaModel.fromJson(jsonData['estoqueMarcaModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idProduto'] = idProduto != 0 ? idProduto : null;
		jsonData['idEstoqueMarca'] = idEstoqueMarca != 0 ? idEstoqueMarca : null;
		jsonData['idEstoqueSabor'] = idEstoqueSabor != 0 ? idEstoqueSabor : null;
		jsonData['idEstoqueTamanho'] = idEstoqueTamanho != 0 ? idEstoqueTamanho : null;
		jsonData['idEstoqueCor'] = idEstoqueCor != 0 ? idEstoqueCor : null;
		jsonData['codigo'] = codigo;
		jsonData['quantidade'] = quantidade;
		jsonData['produtoModel'] = produtoModel?.toJson;
		jsonData['estoqueCorModel'] = estoqueCorModel?.toJson;
		jsonData['estoqueTamanhoModel'] = estoqueTamanhoModel?.toJson;
		jsonData['estoqueSaborModel'] = estoqueSaborModel?.toJson;
		jsonData['estoqueMarcaModel'] = estoqueMarcaModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idProduto = plutoRow.cells['idProduto']?.value;
		idEstoqueMarca = plutoRow.cells['idEstoqueMarca']?.value;
		idEstoqueSabor = plutoRow.cells['idEstoqueSabor']?.value;
		idEstoqueTamanho = plutoRow.cells['idEstoqueTamanho']?.value;
		idEstoqueCor = plutoRow.cells['idEstoqueCor']?.value;
		codigo = plutoRow.cells['codigo']?.value;
		quantidade = plutoRow.cells['quantidade']?.value?.toDouble();
		produtoModel = ProdutoModel();
		produtoModel?.nome = plutoRow.cells['produtoModel']?.value;
		estoqueCorModel = EstoqueCorModel();
		estoqueCorModel?.nome = plutoRow.cells['estoqueCorModel']?.value;
		estoqueTamanhoModel = EstoqueTamanhoModel();
		estoqueTamanhoModel?.nome = plutoRow.cells['estoqueTamanhoModel']?.value;
		estoqueSaborModel = EstoqueSaborModel();
		estoqueSaborModel?.nome = plutoRow.cells['estoqueSaborModel']?.value;
		estoqueMarcaModel = EstoqueMarcaModel();
		estoqueMarcaModel?.nome = plutoRow.cells['estoqueMarcaModel']?.value;
	}	

	EstoqueGradeModel clone() {
		return EstoqueGradeModel(
			id: id,
			idProduto: idProduto,
			idEstoqueMarca: idEstoqueMarca,
			idEstoqueSabor: idEstoqueSabor,
			idEstoqueTamanho: idEstoqueTamanho,
			idEstoqueCor: idEstoqueCor,
			codigo: codigo,
			quantidade: quantidade,
		);			
	}

	
}